packages/gatsby/src/bootstrap/remove-cache.js

```diff
@@ -0,0 +1,22 @@
+const fs = require(`fs-extra`)
+const path = require(`path`)
+
+const CACHE_DIR = `.cache`
+const EXCLUDE_PATHS = [`gatsby-source-filesystem`]
+
+module.exports = function removeCache(program, additionalExclusions = []) {
+  const base = path.join(program.directory, CACHE_DIR)
+
+  const exclude = EXCLUDE_PATHS.concat(additionalExclusions)
+
+  return fs
+    .readdir(base)
+    .then(files =>
+      files
+        .map(file => path.join(base, file))
+        .filter(file =>
+          exclude.every(excludePath => !file.includes(excludePath))
+        )
+    )
+    .then(safeFiles => Promise.all(safeFiles.map(file => fs.remove(file))))
```

Comment:
How long does this take? Seems like we could do this on a directory level — e.g. exclude directories and then remove every other directory

Reply:
I think it's honestly about the same. I'd need to do debug the performance, and I'm not really sure what the best way is to do that.

On a few runs, here are the average times

|method|user|system|
|:----:|:--:|:-----:|
|`fs.remove`|`0.11s`|`0.27s`|
|`fs.readdir()` then `fs.remove()`|`0.09`|`0.16s`|

Reply:
(Also note there's no way this is statistically significant, haha... it was run on gatsby/www a few times, so not the best test case most likely)